operator/controllers/actionsrunner/wire.go

```diff
@@ -0,0 +1,186 @@
+*/
+
+package actionsrunner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	inlocov1alpha1 "github.com/inloco/kube-actions/operator/api/v1alpha1"
+	"github.com/inloco/kube-actions/operator/controllers/actionsrunner/dot"
+	"github.com/inloco/kube-actions/operator/controllers/actionsrunner/facades"
+	"k8s.io/utils/strings"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+type Wire struct {
+	Context context.Context
+	Log     logr.Logger
+
+	ActionsRunner *inlocov1alpha1.ActionsRunner
+	DotFiles      *dot.Files
+
+	ghFacade  facades.GitHub
+	adoFacade facades.AzureDevOps
+
+	loopClose    chan struct{}
+	loopAck      chan struct{}
+	loopMessages chan facades.Message
+	loopErrors   chan error
+}
+
+func (w *Wire) Init() error {
+	if err := w.initDotFiles(); err != nil {
+		return err
+	}
+
+	w.ghFacade.Context = w.Context
+	if err := w.ghFacade.Init(w.ActionsRunner.Spec.Repository.Owner, w.ActionsRunner.Spec.Repository.Name); err != nil {
+		return err
+	}
+	w.DotFiles.Runner.GitHubUrl = w.ghFacade.Repository.GetGitCommitsURL()
+
+	credential, err := w.ghFacade.GetGitHubTenantCredential(facades.RunnerEventRegister)
+	if err != nil {
+		return err
+	}
+	w.DotFiles.Runner.ServerUrl = credential.GetURL()
+
+	w.adoFacade.Context = w.Context
+	if err := w.adoFacade.Init(credential.GetToken(), credential.GetURL(), w.DotFiles, w.ActionsRunner.Spec.Labels); err != nil {
+		return err
+	}
+
+	if w.loopAck == nil {
+		w.loopAck = make(chan struct{})
+	}
+
+	if w.loopMessages == nil {
+		w.loopMessages = make(chan facades.Message)
+	}
+
+	if w.loopErrors == nil {
+		w.loopErrors = make(chan error)
+	}
+
+	return nil
+}
+
+func (w *Wire) initDotFiles() error {
+	if w.DotFiles != nil {
+		return nil
+	}
+
+	dotRSAParameters, err := dot.NewRSAParameters()
+	if err != nil {
+		return err
+	}
+
+	w.DotFiles = &dot.Files{
+		Runner: dot.Runner{
+			AgentName:  strings.ShortenString(fmt.Sprintf("KA %s %s", w.ActionsRunner.GetNamespace(), w.ActionsRunner.GetName()), 64),
+			PoolId:     1,
+			PoolName:   "Default",
+			WorkFolder: "_work",
+		},
+		Credentials: dot.Credentials{
+			Scheme: "OAuth",
+		},
+		RSAParameters: *dotRSAParameters,
+	}
+	return nil
+}
+
+func (w *Wire) Go(eventSource chan<- event.GenericEvent) (<-chan struct{}, <-chan facades.Message, <-chan error) {
+	if !w.isClosed() {
+		return w.loopAck, w.loopMessages, w.loopErrors
+	}
+
+	w.loopClose = make(chan struct{})
+
+	go func() {
+		genericEvent := event.GenericEvent{
+			Meta: w.ActionsRunner,
+		}
+
+		defer func() {
+			if r := recover(); r != nil {
+				eventSource <- genericEvent
+				w.loopErrors <- fmt.Errorf("%v", r)
+				w.Close()
+			}
+		}()
+
+		if err := w.adoFacade.InitAzureDevOpsTaskAgentSession(); err != nil {
+			panic(err)
+		}
+		defer w.adoFacade.DeinitAzureDevOpsTaskAgentSession()
+
+		for !w.isClosed() {
+			w.Log.Info("Getting Message")
+
+			taMessage, err := w.adoFacade.GetMessage()
+			if err != nil {
+				panic(err)
```

Comment:
recover is cool, but we may want just a return here?

Reply:
this happens inside an anonymous goroutine, I'm not able to return the error and treat it latter. I cloud have it replaced by the 3 lines inside the recover function, but I would repeat that piece of code too much.